### PR TITLE
fix #4569: clang warning about deprecated sprintf usage.

### LINKED ISSUE #4569: Clang warning: deprecated sprintf usage in json_reader (Version: 1.11)
## Issue Description
Clang compiler throws the following warning regarding the usage of `sprintf` function.

```
/Users/ckeshavabs/personal-code/rippled/src/ripple/json/impl/json_reader.cpp:928:5: warning: 'sprintf' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of sprintf(3), it is highly recommended that you use snprintf(3) instead. [-Wdeprecated-declarations]
    sprintf(buffer, "Line %d, Column %d", line, column);
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/stdio.h:188:1: note: 'sprintf' has been explicitly marked deprecated here
__deprecated_msg("This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of sprintf(3), it is highly recommended that you use snprintf(3) instead.")
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/sys/cdefs.h:215:48: note: expanded from macro '__deprecated_msg'
        #define __deprecated_msg(_msg) __attribute__((__deprecated__(_msg)))
```

## Steps to Reproduce
Compiling the latest version of `rippled` codebase throws this warning.

## Environment
Here is the version details of the clang compiler:
```
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I'm not sure if this warning is benign and could be ignored. Putting it out here for reference.

### PR DESCRIPTION
**Pull request by ckeshava**
This fix is intended to remove a warning from the clang compiler. It switches the usage of `sprintf` function into the recommended `snprintf` function.

Fix #4569

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

**Comment by ckeshava** [comment]
Here is the related issue: https://github.com/XRPLF/rippled/issues/4569

File: src/ripple/json/impl/json_reader.cpp [hunk 1]
```diff
@@ -925,7 +924,15 @@ Reader::getLocationLineAndColumn(Location location) const
     int line, column;
     getLocationLineAndColumn(location, line, column);
     char buffer[18 + 16 + 16 + 1];
-    sprintf(buffer, "Line %d, Column %d", line, column);
+    int generatedStrLen =
+        snprintf(buffer, 51, "Line %d, Column %d", line, column);
```
Comment (ckeshava) [inline src/ripple/json/impl/json_reader.cpp:928]: here is the documentation for the `snprintf` function: https://cplusplus.com/reference/cstdio/snprintf/

I'm basing the second parameter off of the previous line. But I don't understand why `buffer` needs to have a length of 50. The contents of `buffer` seem to require much lesser space.

File: src/ripple/json/impl/json_reader.cpp [hunk 2]
```diff
@@ -925,7 +924,15 @@ Reader::getLocationLineAndColumn(Location location) const
...
+
+    if (generatedStrLen < 0)
+    {
+        ripple::LogicError(
+            "json_reader: Exception in "
+            "getLocationLineAndColumn - Could not write into buffer\n");
+    }
     return buffer;
```
[the remark this answers is not in the document]
Comment (ckeshava) [reply src/ripple/json/impl/json_reader.cpp]: okay, thanks for the suggestion! I like this much better!

**Review by HowardHinnant — APPROVED** [review]
Thanks!

**Comment by intelliot** [comment]
for next week: check if seelabs will be able to review this

File: src/ripple/json/impl/json_reader.cpp [hunk 3]
```diff
@@ -924,9 +923,8 @@ Reader::getLocationLineAndColumn(Location location) const
 {
     int line, column;
     getLocationLineAndColumn(location, line, column);
-    char buffer[18 + 16 + 16 + 1];
-    sprintf(buffer, "Line %d, Column %d", line, column);
-    return buffer;
+    return "Line " + std::to_string(line) + ", Column " +
+        std::to_string(column);
```
Comment (seelabs) [inline src/ripple/json/impl/json_reader.cpp:927]: This does multiple allocations. Consider using snprintf instead. Something like: 
```
    int line, column;
    getLocationLineAndColumn(location, line, column);
    std::size_t n = 18 + 16 + 16 + 1;
    char buffer[n];
    snprintf(buffer, n, "Line %d, Column %d", line, column);
    return buffer;
```

**Review by seelabs — CHANGES_REQUESTED** [review]

File: src/ripple/json/impl/json_reader.cpp (hunk 3 above)
Comment (ckeshava) [reply src/ripple/json/impl/json_reader.cpp:927]: Okay. I tried to do this in my first commit. But @a-noni-mousse  had reservations with this approach (https://github.com/XRPLF/rippled/pull/4570#discussion_r1229908127)

Should I still go ahead with the snprintf approach?

Comment (seelabs) [reply src/ripple/json/impl/json_reader.cpp:927]: This isn't a performance critical function, the extra allocations aren't _that_ big a deal. On the other hand, I don't find the `snprintf` dangerous here - it's clearly not going to overwrite the buffer and it's very easy to see that. So the issues I'd weigh are:

* 4 memory allocations vs 1 memory allocation
* a high level implementation vs a low level implementation
* this is not a performance critical function, and is unlikely to ever be performance critical

It's your call, I'll approve this with either interface.

Comment (ckeshava) [reply src/ripple/json/impl/json_reader.cpp:927]: Hmm okay, thanks for the comparison.

I feel the `snprintf` adds complexity. I don't understand why we need to allocate `18 + 16 + 16 + 1` as the buffer size (I mimicked the existing code's behavior)

I'm inclined the favor the `std::string` version for it's simplicity.

**Review by seelabs — APPROVED** [review]

**Comment by ckeshava** [comment]
Okay, I messed up the git commands. I want to sign the first "unverified" commit. But I have already merged the latest develop into my feature branch.

I did a rebase and then `exec git commit --amend --no-edit -S` to sign my first commit, but that seems to pull it many unrelated commits into this PR :/

**Comment by intelliot** [comment]
That's because you rebased all of the subsequent commits. Let's meet up this week and try to fix this :)

**Comment by intelliot** [comment]
@ckeshava : can you resolve the conflict?

**Comment by ckeshava** [comment]
apologies, I made a mess out of the commit history.  I have created a cleaner PR here: https://github.com/XRPLF/rippled/pull/4747

**Pull request closed without merging**